builtin/logical/database/dbs/db.go

```diff
@@ -0,0 +1,91 @@
+package dbs
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	postgreSQLTypeName = "postgres"
+	cassandraTypeName  = "cassandra"
+)
+
+var (
+	ErrUnsupportedDatabaseType = errors.New("Unsupported database type")
+)
+
+func Factory(conf *DatabaseConfig) (DatabaseType, error) {
+	switch conf.DatabaseType {
+	case postgreSQLTypeName:
+		var details *sqlConnectionDetails
+		err := mapstructure.Decode(conf.ConnectionDetails, &details)
+		if err != nil {
+			return nil, err
+		}
+
+		connProducer := &sqlConnectionProducer{
+			config:      conf,
+			connDetails: details,
+		}
+
+		credsProducer := &sqlCredentialsProducer{
+			displayNameLen: 23,
+			usernameLen:    63,
+		}
+
+		return &PostgreSQL{
+			ConnectionProducer:  connProducer,
+			CredentialsProducer: credsProducer,
+		}, nil
+
+	case cassandraTypeName:
+		var details *cassandraConnectionDetails
+		err := mapstructure.Decode(conf.ConnectionDetails, &details)
+		if err != nil {
+			return nil, err
+		}
+
+		connProducer := &cassandraConnectionProducer{
+			config:      conf,
+			connDetails: details,
+		}
+
+		credsProducer := &cassandraCredentialsProducer{}
+
+		return &Cassandra{
+			ConnectionProducer:  connProducer,
+			CredentialsProducer: credsProducer,
+		}, nil
+	}
+
+	return nil, ErrUnsupportedDatabaseType
+}
+
+type DatabaseType interface {
+	Type() string
+	CreateUser(createStmt, rollbackStmt, username, password, expiration string) error
```

Comment:
I think the statements should not be passed into these functions; rather they should be intrinsic to the role. A database type really is probably a combination of a connection and a config (where credentials fits in is debatable but fine here for now).

To me it feels like the Create/Renew/Revoke probably belong on the role interface *because* that's where the statements are. They look up the associated DB (or the one requested by the user), get a connection, then run those functions. The super downside there is that when they get the connection they have to figure out what type it is (although the database type is part of the interface so could work for a switch).

Another possibility might be to have CreateUser/RenewUser/RevokeUser in the Database type but to instead pass in a Role interface. That role interface has functions to return the appropriate SQL (e.g. CreateStatements()) which can return our defaults if nothing has been set by the user. In addition, since there is a CredentialsProvider in the database already the username/password/expiration seem unnecessary).

I'm not sure what the right answer is currently, I just know that passing in the statements this way feels icky to me :-)